### PR TITLE
DAT- 59545

### PR DESCRIPTION
**Pull request by guyshemer**
add support for orphan adoption when father exists in es and not in local timbermill instance

### CODE REVIEW

File: timbermill-server/src/test/java/com/datorama/oss/timbermill/TimberLogTest.java [hunk 1]
```diff
@@ -41,19 +56,58 @@ synchronized static void waitForTask(String taskId, TaskStatus status, Elasticse
 		waitForCallable(callable);
 	}
 
+	public static void waitForTaskPredicate(String taskId, Predicate<Task> predicate, long timeout, TimeUnit unit) {
```
Comment (edengal) [inline timbermill-server/src/test/java/com/datorama/oss/timbermill/TimberLogTest.java:59]: package private

Comment (edengal) [reply timbermill-server/src/test/java/com/datorama/oss/timbermill/TimberLogTest.java:59]: make sure all usage of this is only 1 minute, no need for more

File: timbermill-local/src/main/java/com/datorama/oss/timbermill/TaskIndexer.java [hunk 2]
```diff
@@ -229,24 +218,41 @@ public static void logErrorInEventsMap(Map<String, List<Event>> eventsMap, Strin
         }
     }
 
-    private void updateAdoptedOrphans(Map<String, List<Event>> eventsMap, String parentTaskId) {
-        Queue<AdoptedEvent> adoptedEvents = parentIdTORootOrphansEventsCache.getIfPresent(parentTaskId);
+    public static void handleAdoptions(ElasticsearchClient es, int orphansFetchPeriodMinutes, int daysRotation) {
```
[the remark this answers is not in the document]
Comment (DvirKatz) [reply timbermill-local/src/main/java/com/datorama/oss/timbermill/TaskIndexer.java:221]: +1

File: timbermill-local/src/main/java/com/datorama/oss/timbermill/cron/OrphansAdoptionJob.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+import com.datorama.oss.timbermill.TaskIndexer;
+import com.datorama.oss.timbermill.common.ElasticsearchUtil;
+
+public class OrphansAdoptionJob implements Job {
+	private static final Logger LOG = LoggerFactory.getLogger(OrphansAdoptionJob.class);
+	private static final AtomicBoolean currentlyRunning = new AtomicBoolean(false);
+
+	@Override public void execute(JobExecutionContext context) {
+
+		if (currentlyRunning.compareAndSet(false, true)) {
```
Comment (DvirKatz) [inline timbermill-local/src/main/java/com/datorama/oss/timbermill/cron/OrphansAdoptionJob.java:20]: consider using @DisallowConcurrentExecution instead

Comment (guyshemer) [reply timbermill-local/src/main/java/com/datorama/oss/timbermill/cron/OrphansAdoptionJob.java:20]: nice. I wasn't familiar with it

File: timbermill-local/src/main/java/com/datorama/oss/timbermill/unit/Task.java [hunk 4]
```diff
@@ -28,6 +28,7 @@
 	private static final Logger LOG = LoggerFactory.getLogger(Task.class);
 	private static final String OLD_EVENT_ID_DELIMITER = "_";
 	private static final String TIMBERMILL_SUFFIX = "_timbermill2";
+	private static final int RETRIES_ON_CONFLICT = 1;
```
Comment (edengal) [inline timbermill-local/src/main/java/com/datorama/oss/timbermill/unit/Task.java:31]: Let's put 3

**Pull request merged**